lib/src/collection/lru_map.dart

```diff
@@ -0,0 +1,303 @@
+   * [maximumSize].
+   */
+  int maximumSize;
+}
+
+/**
+ * Simple implementation of a linked-list entry that contains a [key] and
+ * [value].
+ */
+class _LinkedEntry<K, V> {
+  K key;
+  V value;
+
+  _LinkedEntry<K, V> next;
+  _LinkedEntry<K, V> previous;
+
+  _LinkedEntry([this.key, this.value]);
+}
+
+/**
+ * A linked hash-table based implementation of [LruMap].
+ */
+class LinkedLruHashMap<K, V> implements LruMap<K, V> {
+  static const _DEFAULT_MAXIMUM_SIZE = 100;
+
+  final Map<K, _LinkedEntry<K, V>> _entries;
+
+  int _maximumSize;
+
+  _LinkedEntry<K, V> _head;
+  _LinkedEntry<K, V> _tail;
+
+  /**
+   * Create a new LinkedLruHashMap with a [maximumSize].
+   */
+  factory LinkedLruHashMap({int maximumSize}) =>
+      new LinkedLruHashMap._fromMap(new HashMap<K, _LinkedEntry<K, V>>(),
+          maximumSize: maximumSize);
+
+  LinkedLruHashMap._fromMap(
+      this._entries, {
+      int maximumSize})
+          // This pattern is used instead of a default value because we want to
+          // be able to respect null values coming in from MapCache.lru.
+          : _maximumSize = firstNonNull(maximumSize, _DEFAULT_MAXIMUM_SIZE);
+
+  /**
+   * Adds all key-value pairs of [other] to this map.
+   *
+   * The operation is equivalent to doing this[key] = value for each key and
+   * associated value in other. It iterates over other, which must therefore not
+   * change during the iteration.
+   *
+   * If a key of [other] is already in this map, its value is overwritten. If
+   * the number of unique keys is greater than [maximumSize] then the least
+   * recently use keys are evicted. For keys written to by [other], the least
+   * recently user order is determined by [other]'s iteration order.
+   */
+  @override
+  void addAll(Map<K, V> other) => other.forEach((k, v) => this[k] = v);
+
+  @override
+  void clear() {
+    _entries.clear();
+    _head = _tail = null;
+  }
+
+  @override
+  bool containsKey(K key) => _entries.containsKey(key);
+
+  @override
+  bool containsValue(V value) => values.contains(value);
+
+  /**
+   * Applies [action] to each key-value pair of the map in order of MRU to LRU.
+   *
+   * Calling `action` must not add or remove keys from the map.
+   */
+  @override
+  void forEach(void action(K key, V value)) {
+    var head = _head;
+    while (head != null) {
+      action(head.key, head.value);
+      head = head.next;
+    }
+  }
+
+  @override
+  int get length => _entries.length;
+
+  @override
+  bool get isEmpty => _entries.isEmpty;
+
+  @override
+  bool get isNotEmpty => _entries.isNotEmpty;
+
+  /**
+   * Creates an [Iterable] around the entries of the map.
+   */
+  Iterable<_LinkedEntry<K, V>> _iterable() {
+    return new GeneratingIterable<_LinkedEntry<K, V>>(
+        () => _head, (n) => n.next);
+  }
+
+  /**
+   * The keys of [this] - in order of MRU to LRU.
+   *
+   * The returned iterable does *not* have efficient `length` or `contains`.
```

Comment:
Interesting... it definitely could with a pretty simple subclass of `DelegatingIterable`. Maybe a todo?